src/browser/transport/web_socket_connection.rs

```diff
@@ -1,45 +1,63 @@
 use std::sync::mpsc;
-use std::sync::Mutex;
+use std::sync::Arc;
 
 use anyhow::Result;
-use log::{debug, info, trace, warn};
-use websocket::client::sync::Client;
-use websocket::stream::sync::TcpStream;
-use websocket::url::Url;
-use websocket::WebSocketError;
-use websocket::{ClientBuilder, OwnedMessage};
+use futures::{
+    stream::{SplitSink, SplitStream},
+    SinkExt,
+};
+use futures_util::stream::StreamExt;
+use log::*;
+use tokio::{net::TcpStream, runtime::Runtime, sync::Mutex};
+use tokio_tungstenite::{connect_async, MaybeTlsStream, WebSocketStream};
+use tungstenite::client::IntoClientRequest;
+use url::Url;
 
 use crate::types::{parse_raw_message, Message};
 
+#[derive(Debug)]
 pub struct WebSocketConnection {
-    sender: Mutex<websocket::sender::Writer<TcpStream>>,
+    sender: Arc<Mutex<SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, tungstenite::Message>>>,
+    runtime: Arc<Runtime>,
     process_id: Option<u32>,
 }
 
-// TODO websocket::sender::Writer is not :Debug...
-impl std::fmt::Debug for WebSocketConnection {
-    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
-        write!(f, "WebSocketConnection {{}}")
-    }
-}
-
 impl WebSocketConnection {
     pub fn new(
         ws_url: &Url,
         process_id: Option<u32>,
         messages_tx: mpsc::Sender<Message>,
     ) -> Result<Self> {
-        let connection = Self::websocket_connection(ws_url)?;
-        let (websocket_receiver, sender) = connection.split()?;
+        let runtime = Arc::new(
+            tokio::runtime::Builder::new_current_thread()
+                .enable_all()
+                .build()?,
+        );
+
+        let connection: Result<_> = runtime.block_on(async {
+            let (stream, _) = connect_async(ws_url.to_string().into_client_request()?).await?;
 
+            debug!("Successfully connected to WebSocket: {}", ws_url);
+
+            Ok(stream)
+        });
+
+        let (sender, recv) = connection?.split();
```

Comment:
does this have any performance or memory impact as we are using a async runtime and then blocking on the result

Reply:
Apparently yes. This runtime object has some memory overhead since it contains schedulers, handles, etc. for a Tokio runtime. However, the execution speed shouldn't be too affected, but I also didn't run benchmarks. This runtime is created in the minimal invasive approach possible: by not spawning new threads but instead scheduling futures in the current thread. This "inlines" the web socket code whenever an event occurs. 

This is not perfect, but the only way to not rewrite the whole interface as async and at the same time mitigate the security vulnerability. I would assume that the memory and execution duration overhead is not notable at all to the end user, but again I did not run benchmarks.

Reply:
> Apparently yes. This runtime object has some memory overhead since it contains schedulers, handles, etc. for a Tokio runtime. However, the execution speed shouldn't be too affected, but I also didn't run benchmarks. This runtime is created in the minimal invasive approach possible: by not spawning new threads but instead scheduling futures in the current thread. This "inlines" the web socket code whenever an event occurs.
> 
> This is not perfect, but the only way to not rewrite the whole interface as async and at the same time mitigate the security vulnerability. I would assume that the memory and execution duration overhead is not notable at all to the end user, but again I did not run benchmarks.

Thank you for explaining all of this and yes this is a good enough solution for now, great job